lib/ts/MemArena.cc

```diff
@@ -0,0 +1,232 @@
+/** @file
+
+    MemArena memory allocator. Chunks of memory are allocated, frozen into generations and
+     thawed away when unused.
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include "MemArena.h"
+#include <ts/ink_memory.h>
+#include <ts/ink_assert.h>
+
+using namespace ts;
+
+inline void
+MemArena::Block::init(size_t n)
+{
+  size      = n;
+  allocated = 0;
+  next      = nullptr;
+}
+
+inline char *
+MemArena::Block::data()
+{
+  return reinterpret_cast<char *>(this + sizeof(Block));
+}
+
+/**
+    Allocates a new internal block of memory. If there are no existing blocks, this becomes the head of the
+     @allocations ll. If there are existing allocations, the new block is inserted in the allocations list.
+     If @back == true, the new block is appended and becomes the @current internal block.
+        @back == false, the new block is pushed into the generation but @current doesn't change.
+  */
+inline MemArena::Block *
+MemArena::newInternalBlock(size_t n, bool back)
+{
+  // Adjust to the nearest power of two. Works for 32 bit values which is ok as we are unlikely to
+  //  allocate over 4GB in this class. Allocate Block header and actual underlying memory together for locality. 
+  //  -16 to account for malloc/free headers.
+  static constexpr size_t free_space_per_page = DEFAULT_PAGE_SIZE - sizeof(Block) - 16;
+
+  void *tmp;
+  if( n <= free_space_per_page) { // will fit within one page, just allocate. 
+    tmp = ats_malloc(n + sizeof(Block));
+  } else {
+    size_t t = n;
+    t--;
+    t |= t >> 1;
+    t |= t >> 2;
+    t |= t >> 4;
+    t |= t >> 8;
+    t |= t >> 16;
```

Comment:
It seems to me that adding `t |= t >> 32;` would work to cover the 64 bit case. You can see in the pattern that it's doubling the number of leading 1 bits each time, so one more step is enough.